include/nlohmann/detail/conversions/to_json.hpp

```diff
@@ -51,6 +51,16 @@ struct external_constructor<value_t::string>
         j.m_value = std::move(s);
         j.assert_invariant();
     }
+
+	template<typename BasicJsonType, typename CompatibleStringType,
+		     enable_if_t<not std::is_same<CompatibleStringType, typename BasicJsonType::string_t>::value,
+		                 int> = 0>
```

Comment:
This SFINAE check is already done in the `to_json` method, maybe we can only do it once:

- Adding a `to_json(const BasicJson::string_t&)` overload (only `BasicJson::string_t&&` is handled right now)
- Then, in the `to_json` overload where the `std::is_constructible<...>` check is performed, construct a `string_t` and move it to the `external_constructor`.

Additionally, I'd rather keep `external_constructor` function count low (Maybe the `array_t` specialization could get cleaned a bit).

This is mostly a style preference though.